File: 150-evaluate-reverse-polish-notation/150-Evaluate-Reverse-Polish-Notation.java
import java.util.Stack;
class Solution {
    public static int eval(int n1, int n2, String op){
        return switch(op){
            case "+"->n1+n2;
            case "-"->n1-n2;
            case "*"->n1*n2;
            case "/"->n1/n2;
            default->throw new IllegalArgumentException("Unknown operator: " +op);
        };
    }
    public int evalRPN(String[] tokens) {
        Stack<Integer> stack = new Stack<>();
        HashSet<String> operators = new HashSet<>(Set.of("+","-","*","/"));
        for(String x:tokens){
            if(!operators.contains(x)){
                stack.push(Integer.parseInt(x));
            }else{
                int n2 = stack.pop();
                int n1 = stack.pop();
                stack.push(eval(n1,n2,x));
            }
        }
        return stack.peek();
    }
}

File: 150-evaluate-reverse-polish-notation/150-Evaluate-Reverse-Polish-Notation.py
class Solution:
    def evalRPN(self, tokens):
        def eval(num1, num2, op):
            match op:
                case "+":
                    return num1 + num2
                case "-":
                    return num1 - num2
                case "*":
                    return num1 * num2
                case "/":
                    return int(num1/num2)
        
        stack = []
        num1 = 0
        num2 = 0
        for i in tokens:
            if i not in {"+","-","*","/"}:
                stack.append(int(i))
            else:
                num2 = stack.pop()
                num1 = stack.pop()
                stack.append(eval(num1,num2,i))

        return stack[-1]
